### PR TITLE
Language Server Protocol Support for Brackets

### PR DESCRIPTION
**Pull request by jhasubhash**
This covers the first cut of Language Server Protocol Support for Brackets.

**Basic core implementation to:**

- Initiate/Stop a Language Server
- Interact with Language Server
  - Completions
  - Signature Help
  - Diagnostics
  - Jump To Dec/Decl/Impl
  - Symbols
  - References
  - Synchronization Events (Text Document didOpen, didChange etc.)
  - Custom Request/Notification (so the developer can use commands not exposed in the initial implementation)
- Bidirectional Request/Notification system for communication between Brackets and Language Server
- Abstraction of LSP from Brackets (as far as possible)
- Creation of Language Client that works as the handle to Language Server from Brackets side
- Support for all communication types: IPC, Pipe, Socket and Standard I/O

**TODO:**
- [x] Move vscode-languageserver-protocol to Thirdparty
- [x] Tooling Manager
  - [x] CodeHints
  - [x] ParameterHints
  - [x] Diagnostics
  - [x] Jump to Def/Decl/Impl
  - [x] Priority
  - [x] Translate brackets events to LSP events
  - [x] Default handlers for all the above Tooling Interfaces
- [x] Project level event handling
  - [x] Project Switch
  - [x] Reload with Extensions
- [x] Grunt modifications for build generation
- [x] UI
- [x] Tests

### This feature can be experienced with the reference Language Client implementation for PHP here : https://github.com/adobe/brackets/pull/14671

### CODE REVIEW

File: src/lsp/ClientTemplate.js [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ *
+ */
+
+define(function (require, exports, module) {
+    "use strict";
+    
+    class Client{
```
Comment (ingorichter) [inline src/lsp/ClientTemplate.js]:
```suggestion
    class Client {
```
Please ensure that the formatting is consistent across the files, e.g. white space between text and `{` as an example.

File: src/lsp/LSPInterface.js [hunk 2]
```diff
@@ -0,0 +1,150 @@
+
```
Comment (ingorichter) [inline src/lsp/LSPInterface.js]: Nit: obsolete empty line

File: src/lsp/LSPInterface.js [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+    /**
+     * Receive any response/notification from the LSP server
+     * @param {Object}  event  - event received from node
+     * @param {Object}  msgObj - json containing - {
+     *                          serverName - name of the LSP server which sent the message
+     *                          param - Content of the message
+     */
+    function receiveMessageFromServer(event, msgObj) {
+        
+        if(msgObj.param.method && _callbackList[msgObj.serverName+msgObj.param.method]){
```
Comment (ingorichter) [inline src/lsp/LSPInterface.js]: I would assert that `msgObj` is defined and param exists before checkin for `method`. Same applies for accessing `_callbackList`. There is a some potential to create a variable and assign the constructed key `msgObj.serverName+msgObj.param`.

File: src/lsp/LSPInterface.js [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ *
+ */
+
+ define(function (require, exports, module) {
```
Comment (ingorichter) [inline src/lsp/LSPInterface.js]: I suggest to write unit tests for all this new functionality to ensure that edge cases and error conditions are properly excercises

File: src/lsp/ParameterHintManager.js [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+    }
+
+    /**
+     * Test if a function hint is being displayed.
+     *
+     * @return {boolean} - true if a function hint is being displayed, false
+     * otherwise.
+     */
+    function isHintDisplayed() {
+        return hintState.visible === true;
```
Comment (ingorichter) [inline src/lsp/ParameterHintManager.js]:
```suggestion
        return hintState.visible;
```
should be enough ;-)

File: src/lsp/ParameterHintManager.js [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+            $hintContent.append(_.escape(hints));
+        }
+    }
+
+    /**
+     * Dismiss the function hint.
+     *
+     */
+    function dismissHint() {
+        if (hintState.visible) {
```
Comment (ingorichter) [inline src/lsp/ParameterHintManager.js]: nit: you could call `isHintDisplayed` instead

**Review by ingorichter — COMMENTED** [review]
I can’t test it myself right now, but I wanted to leave a couple of comments before I’ll look into the functional details. Besides the feature implementation I recommend to run the linter and ensure that the formatting is consistent with all the remaining brackets sources (there might be inconsistencies too). Since this is a lot of new code we should ensure that it’s properly formatted and documented. Adding test cases is crucial for this sort of functionality. Adding a sample LSP client would help to demonstrate the use of the API and helps with the testing of this great new feature.

File: src/languageTools/LanguageClient/ProtocolAdapter.js [hunk 7]
```diff
@@ -0,0 +1,337 @@
...
+}
+
+function processNotification(connection, message) {
+    _constructParamsAndRelay(connection, message.type, message.params);
+}
+
+function attachOnNotificationHandlers(connection, handler) {
+    function _callbackFactory(type) {
+        switch (type) {
+        case ToolingInfo.protocol.ShowMessageNotification.type:
```
Comment (niteskum) [inline src/languageTools/LanguageClient/ProtocolAdapter.js]: ToolingInfo Object does not have protocol properties?

Comment (shubhsnov) [reply src/languageTools/LanguageClient/ProtocolAdapter.js]: Yeah it doesn't.

File: src/languageTools/LanguageTools.js [hunk 8]
```diff
@@ -0,0 +1,366 @@
...
+        self.stop().done(function() {
+            self.start();
+        });
+        //TODO : the case where server supports change of workspace folders
+        //TODO : Reload with extensions case
+    };
+
+    function _attachEventHandlers(client) {
+        EditorManager.on("activeEditorChange.language-tools", client._handleActiveEditorChange.bind(client));
+        ProjectManager.on("projectOpen", client._handleProjectOpen.bind(client));
```
Comment (niteskum) [inline src/languageTools/LanguageTools.js]: This will work only if file or Project is opened after extension is loaded.
Need a check here that if there is any editor is active. _handleActiveEditor should be called and then attach event handlers.

File: src/languageTools/LanguageTools.js [hunk 9]
```diff
@@ -0,0 +1,366 @@
...
+    }
+
+    function intiateToolingService(clientName, clientFilePath, languages) {
+        var result = $.Deferred();
+
+        MessageHandler.initiateLanguageClient(clientName, clientFilePath)
```
Comment (niteskum) [inline src/languageTools/LanguageTools.js]: By default Server Process spawned by a bracket window should be stopped whenever window is closed.
but if Extension developer does not want this default behavior , he should be able to override this.

**Comment by swmitra** [comment]
Rename all the files starting with lowercase.

**Comment by shubhsnov** [comment]
> Is this check necessary? I think we should just have not null check. if Params is not null, It should be in the proper format

No that won't work. I've tried to not use the LSP convention for params. The params are mostly based on the Brackets Convention. We convert the params to LSP format in ProtocolAdapter.js. The check for the format is only to be used for debugging purposes when we want to pass LSP format hints from the brackets side. It is advised that we use normal Brackets convention for passing params. I'll document the params soon. For debugging, you can use the format specified in MessageHandler.js.

File: src/languageTools/LanguageClientWrapper.js [hunk 10]
```diff
@@ -0,0 +1,377 @@
...
+    };
+
+    /**
+        textDocument requests
+    */
+    //completion
+    LanguageClientWrapper.prototype.requestHints = function (params) {
+        return this._request(ToolingInfo.FEATURES.CODE_HINTS, params);
+    };
+
```
Comment (niteskum) [inline src/languageTools/LanguageClientWrapper.js]: Add wrapper method to resolveHints(method: ‘completionItem/resolve’) also.
https://microsoft.github.io/language-server-protocol/specification#completionItem_resolve

Comment (niteskum) [reply src/languageTools/LanguageClientWrapper.js]: ResolveHints will be called from Provider insertHints function.

**Comment by shubhsnov** [comment]
@adobe/open-source-office Please take care of the CLA check failure.

**Comment by stevengill** [comment]
Hey @shubhsnov is the author of this PR (@subhashjha333) an Adobe employee?

P.S. This is steve from the open source office.

**Comment by swmitra** [comment]
Yes @stevengill, @subhashjha333 is an active Adobe employee.

**Comment by stevengill** [comment]
@swmitra I have sent @subhashjha333 an invite to this adobe org on github. He needs to go to github.com/adobe and accept the invite at the top of that page. Once that is done, close and re-open this PR and the check should pass. 

I'd also recommend @subhashjha333 add `@Adobe` to his github profile under the company field. 

If any other adobe employee needs the CLA check, all they have to do is get added to this org. They can add themselves by following the instructions at https://git.corp.adobe.com/OpenSourceAdvisoryBoard/handbook/blob/master/GitHub-Adobe-Org-Management.md#request-access-to-our-adobe-github-org. This way we can get their adobe ID to github ID mapping. This request will send an invitation to this org which they will accept. After acceptance, close and re-open the PR to trigger the check again.

File: src/languageTools/LanguageTools.js [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+        ProjectManager.on(_withNamespace(BRACKETS_EVENTS_NAMES.PROJECT_CLOSE_EVENT), _eventHandler); //(event, directory)
+        DocumentManager.on(_withNamespace(BRACKETS_EVENTS_NAMES.DOCUMENT_DIRTY_EVENT), _eventHandler); //(event, document)
+        DocumentModule.on(_withNamespace(BRACKETS_EVENTS_NAMES.DOCUMENT_CHANGE_EVENT), _eventHandler); //(event, document, changeList)
+        DocumentManager.on(_withNamespace(BRACKETS_EVENTS_NAMES.FILE_RENAME_EVENT), _eventHandler); //(event, oldName, newName)
+        ProjectManager.on(_withNamespace(BRACKETS_EVENTS_NAMES.BEFORE_APP_CLOSE), _eventHandler); //(event, oldName, newName)
+    }
+
+    _attachEventHandlers();
+
+    function intiateToolingService(clientName, clientFilePath, languages) {
```
Comment (niteskum) [inline src/languageTools/LanguageTools.js]: typo in function Name
initiateToolingService

**Comment by vickramdhawal** [comment]
@shubhsnov Can you add documents related ti this PR to the wiki since its a fundamental feature that is being introduced for the longer run.
It will be helpful is design documents and other implementation details are present on the wiki.

File: src/languageTools/DefaultProviders.js [hunk 12]
```diff
@@ -0,0 +1,362 @@
...
+            errors.push(err);
+        });
+
+        this._results = {
+            errors: errors
+        };
+    };
+
+    LintingProvider.prototype.getInspectionResults = function () {
+        return this._results;
```
[the remark this answers is not in the document]
Comment (niteskum) [reply src/languageTools/DefaultProviders.js]: ignore this , as it is fixed

File: src/languageTools/DefaultProviders.js [hunk 13]
```diff
@@ -0,0 +1,362 @@
...
+                    line: obj.range.start.line,
+                    ch: obj.range.start.character
+                },
+                message: obj.message,
+                type: (obj.severity === 1 ? CodeInspection.Type.ERROR : (obj.severity === 2 ? CodeInspection.Type.WARNING : CodeInspection.Type.META))
+            };
+            errors.push(err);
+        });
+
+        this._results = {
```
[the remark this answers is not in the document]
Comment (niteskum) [reply src/languageTools/DefaultProviders.js]: fixed

File: src/languageTools/DefaultProviders.js [hunk 14]
```diff
@@ -0,0 +1,379 @@
...
+        this._results = {};
+    }
+
+    /**
+     * Publish the diagnostics information related to current document
+     * @param   {Object} msgObj - json object containg information associated with 'textDocument/publishDiagnostics' notification from server
+     */
+    LintingProvider.prototype.setInspectionResults = function (msgObj) {
+        let diagnostics = msgObj.diagnostics,
+            filePath = PathConverters.uriToPath(msgObj.uri),
```
[the remark this answers is not in the document]
Comment (niteskum) [reply src/languageTools/DefaultProviders.js]: fixed

File: src/languageTools/LanguageClientWrapper.js (hunk 10 above)
Comment (niteskum) [reply src/languageTools/LanguageClientWrapper.js]: ignore this

File: src/languageTools/LanguageClientWrapper.js [hunk 15]
```diff
@@ -0,0 +1,620 @@
...
+    //shutdown
+    LanguageClientWrapper.prototype.stop = function () {
+        return this._stopClient();
+    };
+
+    //restart
+    LanguageClientWrapper.prototype.restart = function (params) {
+        var self = this;
+        return this.stop().done(function () {
```
Comment (niteskum) [inline src/languageTools/LanguageClientWrapper.js]: Check this once , restart not resolving with start result

File: src/languageTools/LanguageClientWrapper.js [hunk 16]
```diff
@@ -0,0 +1,620 @@
...
+    };
+
+    /**
+        Requests
+    */
+    //start
+    LanguageClientWrapper.prototype.start = function (params) {
+        params = validateRequestParams(ToolingInfo.LANGUAGE_SERVICE.START, params);
+        if (params) {
+            return this._startClient(params);
```
Comment (niteskum) [inline src/languageTools/LanguageClientWrapper.js]: if we set serverCapabilites to this object here itself when start resolves. extension developer will not have to set explicitly to client.??

**Review by niteskum — CHANGES_REQUESTED** [review]

File: src/languageTools/LanguageClientWrapper.js (hunk 15 above)
Comment (shubhsnov) [reply src/languageTools/LanguageClientWrapper.js]: fixed

File: src/languageTools/LanguageClientWrapper.js (hunk 16 above)
Comment (shubhsnov) [reply src/languageTools/LanguageClientWrapper.js]: fixed

File: src/languageTools/LanguageClientWrapper.js [hunk 17]
```diff
@@ -0,0 +1,620 @@
...
+            },
+            {
+                methodName: ToolingInfo.LANGUAGE_SERVICE.NOTIFY,
+                methodHandle: this._onNotificationDelegator.bind(this)
+            }
+        ]);
+
+        //create function interfaces
+        this._startClient = this._domainInterface.createInterface(ToolingInfo.LANGUAGE_SERVICE.START, true);
+        this._stopClient = this._domainInterface.createInterface(ToolingInfo.LANGUAGE_SERVICE.STOP);
```
[the remark this answers is not in the document]
Comment (shubhsnov) [reply src/languageTools/LanguageClientWrapper.js]: fixed

File: src/languageTools/LanguageClient/package.json [hunk 18]
```diff
@@ -0,0 +1,19 @@
+{
+  "name": "langaugeclient",
```
Comment (petetnt) [inline src/languageTools/LanguageClient/package.json]: typo

```suggestion
  "name": "languageclient",
```

or maybe something like

```suggestion
  "name": "brackets-language-client",
```

File: src/languageTools/LanguageClient/package.json [hunk 19]
```diff
@@ -0,0 +1,19 @@
...
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1"
+  },
+  "keywords": [
+    "LSP",
+    "LanguageClient",
+    "Brackets"
+  ],
+  "author": "Adobe",
+  "license": "ISC",
```
Comment (petetnt) [inline src/languageTools/LanguageClient/package.json]:
```suggestion
  "license": "MIT",
```

File: src/languageTools/LanguageTools.js [hunk 20]
```diff
@@ -0,0 +1,121 @@
...
+
+        ClientLoader.syncPrefsWithDomain(languageToolsPrefs);
+    });
+
+    function registerLanguageClient(clientName, languageClient) {
+        languageClients.set(clientName, languageClient);
+    }
+
+    function _withNamespace(event) {
+        var retval = event.split(" ")
```
Comment (petetnt) [inline src/languageTools/LanguageTools.js]: Nit: can just `return event.split(" ")`...

File: src/languageTools/PathConverters.js [hunk 21]
```diff
@@ -0,0 +1,82 @@
+/*
```
Comment (petetnt) [inline src/languageTools/PathConverters.js]: This file is the same as `LanguageClient/Utils.js`, should they just use same file?

File: src/languageTools/styles/default_provider_style.css [hunk 22]
```diff
@@ -0,0 +1,134 @@
...
+    white-space: normal;
+    box-sizing: border-box;
+    float: left;
+    clear: left;
+    max-height: 2em;
+    overflow: hidden;
+    text-overflow: ellipsis;
+    line-height: 1em;
+    -webkit-line-clamp: 2;
+    -webkit-box-orient: vertical;
```
Comment (petetnt) [inline src/languageTools/styles/default_provider_style.css]: Deprecated property?

File: src/languageTools/styles/default_provider_style.css [hunk 23]
```diff
@@ -0,0 +1,134 @@
...
+}
+
+.highlight .hint-description {
+    display: block;
+    color: #6495ed !important;
+}
+
+.highlight .hint-doc {
+    display: -webkit-box;
```
Comment (petetnt) [inline src/languageTools/styles/default_provider_style.css]: Should this and `-webkit-box-orient` just be `display: flex` and `flex-direction: column`?

**Review by petetnt — COMMENTED** [review]
Some comments :) haven't worked with Brackets codebase in a while so there are couple of questions, couple of smaller nits and maybe one or two issues. Generally looks good to me otherwise

File: src/extensions/default/JavaScriptCodeHints/ParameterHintsProvider.js [hunk 24]
```diff
@@ -0,0 +1,229 @@
...
+     *      hint has completed. Returns null, if the function hint is already
+     *      displayed or there is no function hint at the cursor.
+     *
+     */
+    JSParameterHintsProvider.prototype._getParameterHint = function (pushExistingHint, hint, functionInfo) {
+        var result = $.Deferred();
+        functionInfo = functionInfo || this.session.getFunctionInfo();
+        if (!functionInfo.inFunctionCall) {
+            this.cleanHintState();
+            result.reject(null);
```
[the remark this answers is not in the document]
Comment (shubhsnov) [reply src/extensions/default/JavaScriptCodeHints/ParameterHintsProvider.js]: Fixed. Thanks for the catch. It can return early. Basically, since we were returning a promise which could be resolved only once, it was still functioning correctly.

File: src/extensions/default/JavaScriptCodeHints/ParameterHintsProvider.js [hunk 25]
```diff
@@ -0,0 +1,229 @@
...
+        var functionInfo = this.session.getFunctionInfo();
+
+        return functionInfo.inFunctionCall;
+    };
+
+    JSParameterHintsProvider.prototype.getParameterHints = function (explicit, onCursorActivity) {
+        var functionInfo = this.session.getFunctionInfo(),
+            result = null;
+        
+        if (onCursorActivity) {
```
[the remark this answers is not in the document]
Comment (shubhsnov) [reply src/extensions/default/JavaScriptCodeHints/ParameterHintsProvider.js]: Done.

File: src/features/PriorityBasedRegistration.js [hunk 26]
```diff
@@ -0,0 +1,139 @@
...
+
+    /**
+     * @private
+     * Remove a code hint provider
+     * @param {!CodeHintProvider} provider Code hint provider to remove
+     * @param {(string|Array.<string>)=} targetLanguageId Optional set of
+     *     language IDs for languages to remove the provider for. Defaults
+     *     to all languages.
+     */
+    RegistrationHandler.prototype.removeProvider = function (provider, targetLanguageId) {
```
[the remark this answers is not in the document]
Comment (shubhsnov) [reply src/features/PriorityBasedRegistration.js]: Actually, the function isn't meant to be private. Fixed the documentation.

File: src/languageTools/DefaultEventHandlers.js [hunk 27]
```diff
@@ -0,0 +1,194 @@
...
+        ProjectManager = require("project/ProjectManager"),
+        PathConverters = require("languageTools/PathConverters");
+
+    function EventPropagationProvider(client) {
+        this.client = client;
+        this.previousProject = "";
+        this.currentProject = ProjectManager.getProjectRoot();
+    }
+
+    EventPropagationProvider.prototype._sendDocumentOpenNotification = function (languageId, document) {
```
[the remark this answers is not in the document]
Comment (shubhsnov) [reply src/languageTools/DefaultEventHandlers.js]: Fixed

File: src/languageTools/DefaultProviders.js [hunk 28]
```diff
@@ -0,0 +1,381 @@
...
+                line: -1,
+                ch: -1
+            },
+            editor = EditorManager.getActiveEditor(),
+            cursor = editor.getCursorPos(),
+            token = $hint.data("token"),
+            txt = null,
+            query = this.query;
+
+        start = {
```
[the remark this answers is not in the document]
Comment (shubhsnov) [reply src/languageTools/DefaultProviders.js]: Fixed

File: src/languageTools/DefaultProviders.js [hunk 29]
```diff
@@ -0,0 +1,381 @@
...
+            end = {
+                line: token.textEdit.range.end.line,
+                ch: token.textEdit.range.end.character
+            };
+        }
+
+        if (editor) {
+            editor.document.replaceRange(txt, start, end);
+        }
+        // Return false to indicate that another hinting session is not needed
```
[the remark this answers is not in the document]
Comment (shubhsnov) [reply src/languageTools/DefaultProviders.js]: It can. ParameterHintsManager actually mirrors CodeHintsManager in this way. hasParameterHints basically just allows us to add any contextual inspection logic to determine if we can even have parameter hints at the location. LSP however is different, the inspection as well as inference happens in the same call, so here we just check the server abilities and return false. For non-LSP based implementation it could still return true.

File: src/languageTools/DefaultProviders.js [hunk 30]
```diff
@@ -0,0 +1,381 @@
...
+        }).fail(function () {
+            $deferredHints.reject();
+        });
+
+        return $deferredHints;
+    };
+
+    /**
+     * Utility function to make the jump
+     * @param   {Object} curPos - target postion fo the cursor after the jump
```
[the remark this answers is not in the document]
Comment (shubhsnov) [reply src/languageTools/DefaultProviders.js]: Fixed

File: src/languageTools/DefaultProviders.js [hunk 31]
```diff
@@ -0,0 +1,381 @@
...
+        return $deferredHints;
+    };
+
+    function LintingProvider() {
+        this._results = {};
+    }
+
+    /**
+     * Publish the diagnostics information related to current document
+     * @param   {Object} msgObj - json object containg information associated with 'textDocument/publishDiagnostics' notification from server
```
[the remark this answers is not in the document]
Comment (shubhsnov) [reply src/languageTools/DefaultProviders.js]: Fixed

File: src/languageTools/DefaultProviders.js [hunk 32]
```diff
@@ -0,0 +1,381 @@
...
+     */
+    LintingProvider.prototype.setInspectionResults = function (msgObj) {
+        let diagnostics = msgObj.diagnostics,
+            filePath = PathConverters.uriToPath(msgObj.uri),
+            errors = [];
+        diagnostics.forEach(obj => {
```
[the remark this answers is not in the document]
Comment (shubhsnov) [reply src/languageTools/DefaultProviders.js]: Fixed

File: src/languageTools/LanguageClient/Connection.js [hunk 33]
```diff
@@ -0,0 +1,134 @@
...
+            Restart: 1
+        },
+        OnError: {
+            Ignore: 0,
+            Stop: 1
+        }
+    };
+
+    function ActionController() {
+        this.retartsTimes = [];
```
[the remark this answers is not in the document]
Comment (shubhsnov) [reply src/languageTools/LanguageClient/Connection.js]: Fixed

File: src/languageTools/LanguageClient/NodeToBracketsInterface.js [hunk 34]
```diff
@@ -0,0 +1,213 @@
...
+
+    NodeToBracketsInterface.prototype.createInterface = function (methodName, respond) {
+        var self = this;
+        return function (params) {
+            var callObject = {
+                method: methodName,
+                params: params
+            };
+
+            var retval = undefined;
```
[the remark this answers is not in the document]
Comment (shubhsnov) [reply src/languageTools/LanguageClient/NodeToBracketsInterface.js]: We can do it to make the return value uniform but essentially any call to Brackets from the Node(Server) side will either be a notification or request. If it's a notification then ideally we would just call the function and not expect a response which is why I am only returning a promise if it's a request.

File: src/languageTools/LanguageClient/package.json (hunk 18 above)
Comment (shubhsnov) [reply src/languageTools/LanguageClient/package.json]: Fixed

File: src/languageTools/LanguageClient/package.json (hunk 19 above)
Comment (shubhsnov) [reply src/languageTools/LanguageClient/package.json]: Fixed

File: src/languageTools/LanguageTools.js (hunk 20 above)
Comment (shubhsnov) [reply src/languageTools/LanguageTools.js]: Fixed

File: src/languageTools/PathConverters.js (hunk 21 above)
Comment (shubhsnov) [reply src/languageTools/PathConverters.js]: LanguageClient here would work in the Node Context while Utils works in the CEF/Brackets context. The function APIs and logic are the same but there are subtle differences internally wherein the former one we use node.js built-ins and browser ones in the later.

File: src/languageTools/styles/default_provider_style.css (hunk 22 above)
Comment (shubhsnov) [reply src/languageTools/styles/default_provider_style.css]: Actually currently we're running on a slightly older versions of CEF 2623 in Win, 2704 on Mac and 2785 in Linux, where the property works fine. We're using this property elsewhere in brackets core as well, so I think we should do the cleanup collectively together as an issue when we upgrade CEF.

File: src/languageTools/styles/default_provider_style.css (hunk 23 above)
Comment (shubhsnov) [reply src/languageTools/styles/default_provider_style.css]: See my comment above.

File: src/languageTools/LanguageClient/LanguageClient.js [hunk 35]
```diff
@@ -0,0 +1,232 @@
...
+    } else {
+        console.error("Invalid options provided for client :", this._clientName);
+    }
+};
+
+LanguageClient.prototype.start = function (params) {
+    var self = this;
+
+    //Check to see if a connection to a language server already exists.
+    if (self._connection) {
```
[the remark this answers is not in the document]
Comment (shubhsnov) [reply src/languageTools/LanguageClient/LanguageClient.js]: I call it "restart" because the function will be invoked when the connection gets an onClose event. In that case, we'll try to automatically restart the server and reestablish the connection internally.

And it's not the function which restarts the server on workspaceChange. That function is in LanguageClientWrapper.js

You can see its usage in DefaultEventHandlers.js in handleProjectOpen.

**Comment by shubhsnov** [comment]
@petetnt Thanks for the review. I've addressed the comments.

**Review by swmitra — APPROVED** [review]
LGTM

**Comment by grire974** [comment]
Wrong place for this comment I'm sure - but I just saw this feature (looks awesome thanks!) can't find documentation anywhere on how to implement it; for example can I install any of these LSP's? 
https://langserver.org/ 

Some of the LSP's in this list (e.g. I'm interested in Java at the moment) have examples on how to install them on VSCode for example; I looked on the brackets wiki & don't see anything there either under Extensions->Languages that makes any mention of this new functionality.

**Comment by shubhsnov** [comment]
@grire974 This is a new feature and we're actively working on making the documentation better and more accessible for the developers. You can look at this [page](https://github.com/adobe/brackets/wiki/Language-Server-Protocol-Support-in-Brackets) to get started.

We've tried to make sure that the Server integration part isn't too far off from how it works in VSCode, so basically, any extension which works for VSCode will work for us and often in the same way in terms of code pertaining to spawning a language server or interacting with it.

Please reach out to us in case you have any queries in the form of issues. Thanks for your interest.

**Comment by grire974** [comment]
Ah ok - that's pretty informative thanks.  Sounds like then to answer my own question that the LSP's I linked to at langserver.org above might not work out of the box but could possibly be a starting point to modify into the brackets format (e.g. main.js etc...)?  I've written a couple of basic brackets extensions before, so this could be a good next project for me!
